From 92a48eb33e00b5ddb4ac23621d52e30dbcdca1be Mon Sep 17 00:00:00 2001
From: Tom Marshall <tdm.code@gmail.com>
Date: Thu, 21 Jun 2018 01:13:19 +0200
Subject: [PATCH 43/50] recovery: Scale logo image if necessary

Some unified devices (like find7) have differing screen sizes and
choose the larger size for resources.  Handle this by scaling the
logo image dynamically rather than adding more specialized logic
to the device tree.

A complete solution would also scale all of the UI elements (back
icon and main menu icons).  But these are not as sensitive to
screen size and scaling them is not currently necessary.

Change-Id: Ibab0cedd3b15b5fcb7e63595b67119dd7c296f6a
---
 screen_ui.cpp | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/screen_ui.cpp b/screen_ui.cpp
index 2ca28911..6e8fb81a 100644
--- a/screen_ui.cpp
+++ b/screen_ui.cpp
@@ -883,7 +883,22 @@ bool ScreenRecoveryUI::Init(const std::string& locale) {
   // Set up the locale info.
   SetLocale(locale);
 
-  LoadBitmap("logo_image", &logo_image);
+  // Load logo and scale it if necessary
+  // Note 2/45 is our standard margin on each side so the maximum image
+  // width is 41/45 of the screen width.
+  GRSurface* image;
+  LoadBitmap("logo_image", &image);
+  if ((int)gr_get_width(image) > gr_fb_width() * 41 / 45) {
+      float scale = (float)gr_fb_width() / (float)gr_get_width(image) *
+                    (41.0f / 45.0f);
+      GRSurface* scaled_image;
+      res_create_scaled_surface(&scaled_image, image, scale, scale);
+      logo_image = scaled_image;
+      res_free_surface(image);
+  } else {
+      logo_image = image;
+  }
+
   LoadBitmap("ic_back", &ic_back);
   LoadBitmap("ic_back_sel", &ic_back_sel);
 
-- 
2.17.1

